<?xml version="1.0" ?>
<launch>
    <param name="/use_sim_time" value="true"/>
    <arg name="bag_root"   value="/home/cschulz/Bagfiles"/>
    <arg name="bag_file_0" value="velodyne_0.bag" /> 
    <arg name="bag_file_1" value="velodyne_1.bag" />
    <arg name="bag_args"   value="--clock --rate 10.0 -k --pause -q" /> <!-- 0.01 0.0005 -->
	<!-- -s 8.0; -s 75.0 -->

    <group ns="cslibs_mapping">
        <param name="node_rate" value="0.0" />

        <!-- data providers -->
        <group ns="velodyne">
          <param name="class"                    value="cslibs_plugins_data::Pointcloud3dProvider_d"/>
          <param name="base_class"               value="cslibs_plugins_data::DataProvider"/>
          <param name="topic"                    value="/velodyne_points"/>
          <param name="queue_size"               value="0"/>
          <param name="rate"                     value="0.0"/>
        </group>

        <!-- 3D mappers -->
        <group ns="map">
          <param name="class"      	 value="cslibs_mapping::mapper::OruNDTOMFGGridMapper3D"/>
<!-- OccupancyGridMapper3D
     OruNDTOMGridMapper3D, 
     OruNDTOMFGGridMapper3D, 
     NDTGridMapper3D_d_array, 
     NDTGridMapper3D_d_kdtree, 
     NDTGridMapper3D_d_map, 
     NDTGridMapper3D_d_umap, 
     NDTGridMapper3D_d_ucmap, 
     NDTGridMapper3D_f_kdtree, 
     NDTGridMapper3D_f_map, 
     NDTGridMapper3D_f_umap, 
     NDTGridMapper3D_f_ucmap, 
     OccupancyNDTGridMapper3D_d_array, 
     OccupancyNDTGridMapper3D_d_kdtree, 
     OccupancyNDTGridMapper3D_d_map, 
     OccupancyNDTGridMapper3D_d_umap, 
     OccupancyNDTGridMapper3D_d_ucmap, 
     OccupancyNDTGridMapper3D_f_array, 
     OccupancyNDTGridMapper3D_f_kdtree, 
     OccupancyNDTGridMapper3D_f_map, 
     OccupancyNDTGridMapper3D_f_umap, 
     OccupancyNDTGridMapper3D_f_ucmap -->
          <param name="base_class" 	 value="cslibs_mapping::mapper::Mapper"/>	
          <rosparam param="data_providers">
                ["velodyne"]
          </rosparam>
          <rosparam param="map_publishers">
                ["oru_ndt_3d"]  <!-- oru_ndt_3d, ndt_3d_pointcloud, octomap_3d -->
          </rosparam>
          <param name="map_frame"	 value="/odom"/>
          <param name="resolution" 	 value="5.0"/>
          <rosparam param="extents">[665.0,610.0,40.0]</rosparam> 
          <!-- geb: [43.7,18.2,3.3],    fr: [292.5,167.0,28.0],    new: [161.0,251.0,33.0] -->
          <rosparam param="min_point">[-300,-600,-12.5]</rosparam>
          <!-- geb: [-20.0,-17.0,-0.5], fr: [-140.0,-267.5,-16.5], new: [-30,-45.0,15] -->

        <param name="iterations" value="100.0"/>
        <param name="ndt_oru_size_x" value="665.0" />
        <!-- geb: 43.7, fr: 292.5, new: 165.0-->
        <param name="ndt_oru_size_y" value="610.0" />
        <!-- geb: 18.2, fr: 167.0, new: 250.0-->
        <param name="ndt_oru_size_z" value="40.0" /> 
        <!-- geb:  3.3, fr:  28.0, new:  60.0-->
        <param name="ndt_oru_cen_x" value="150.0" /> 
        <!-- geb: 11.8, fr:  82.5, new:  82.5-->
        <param name="ndt_oru_cen_y" value="20.0" />  
        <!-- geb: -2.0, fr: -50.0, new: 100.0-->
        <param name="ndt_oru_cen_z" value="0.0" />  
        <!-- geb: 1.15, fr:   8.0, new:  20.0-->
        </group>
<!--
        <group ns="octomap">
          <param name="class"      	 value="cslibs_mapping::mapper::OccupancyGridMapper3D"/>
          <param name="base_class" 	 value="cslibs_mapping::mapper::Mapper"/>	
          <rosparam param="data_providers">
                ["velodyne"]
          </rosparam>
          <rosparam param="map_publishers">
                ["octomap_3d"]
          </rosparam>
          <param name="map_frame"	 value="/odom"/>
          <param name="resolution" 	 value="5.0"/>
        </group>
-->
        <!-- 3D publishers -->
        <group ns="ndt_3d_pointcloud">
          <param name="class"		    	value="cslibs_mapping::publisher::PointcloudPublisher_d"/>
          <param name="base_class"	    	value="cslibs_mapping::publisher::Publisher"/>
          <param name="rate"		    	value="0.0"/>
          <param name="topic"		    	value="/cslibs_mapping/ndt_3d/map/pointcloud"/>
          <param name="occupancy_ndt"		value="true"/>
          <param name="allocate_all"		value="true"/>
        </group>
        <group ns="octomap_3d">
          <param name="class"		    	value="cslibs_mapping::publisher::OctomapPublisher"/>
          <param name="base_class"	    	value="cslibs_mapping::publisher::Publisher"/>
          <param name="rate"		    	value="0.0"/>
          <param name="topic"		    	value="/cslibs_mapping/ndt_3d/map/octomap"/>
        </group>
        <group ns="oru_ndt_3d">
          <param name="class"		    	value="cslibs_mapping::publisher::OruPublisher"/>
          <param name="base_class"	    	value="cslibs_mapping::publisher::Publisher"/>
          <param name="rate"		    	value="0.0"/>
          <param name="topic"		    	value="/cslibs_mapping/ndt_3d/map/oru"/>
          <param name="occupancy_ndt"		value="false"/>
          <param name="allocate_all"		value="true"/>
        </group>
    </group>

    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find cslibs_mapping)/config/rawseeds.rviz" required="true"/>
    <node pkg="rosbag" type="play" name="rosbag" args="$(arg bag_args) $(arg bag_root)/$(arg bag_file_0) $(arg bag_root)/$(arg bag_file_1)" required="true" output="screen"/>

    <node name="cslibs_mapping" pkg="cslibs_mapping" type="cslibs_mapping_node" output="screen" required="true" clear_params="true" launch-prefix=""/> <!--  launch-prefix="heaptrack" | launch-prefix="valgrind" -->
</launch>
